espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/processor/PointerProcessor.java

```diff
@@ -0,0 +1,145 @@
+
+        // 4 if pause return PAUSE action
+        if (subType.equals(PAUSE)) {
+            return processPauseAction(action, inputSource.getType(), id, index);
+        }
+
+
+        // 5-8 check type and return proper action object
+        switch (subType) {
+            case POINTER_DOWN:
+            case POINTER_UP:
+                actionObject = processPointerUpOrDownAction(action, inputSource.getType(), id, index);
+                break;
+            case POINTER_MOVE:
+                actionObject = processPointerMoveAction(action, inputSource.getType(), id, index);
+                break;
+            case POINTER_CANCEL:
+                throw new NotYetImplementedException();
+            default:
+                // Technically unreachable because the 'validKeyTypes' check catches this
+                throw new InvalidArgumentException(String.format("Invalid pointer type %s", subType));
+        }
+
+        actionObject.setPointer(inputSource.getPointerType());
+        return actionObject;
+    }
+
+    /**
+     * Follows the 'process a pointer up or pointer down action' algorithm in 17.2
+     * @param action Action being processed
+     * @param inputSourceType Source type
+     * @param id ID of input source that it's part of
+     * @param index Index within the 'actions' array
+     * @return Processed action object
+     * @throws InvalidArgumentException If failed to process, throw this. Means that args are bad.
+     */
+    public static ActionObject processPointerUpOrDownAction(Action action, InputSourceType inputSourceType, String id, int index) throws InvalidArgumentException {
+        ActionObject actionObject = new ActionObject(id, inputSourceType, action.getType(), index);
+        int button = action.getButton();
+        if (button < 0) {
+            throwArgException(index, id, String.format("property 'button' must be greater than or equal to 0. Found %s", button));
+        }
+        actionObject.setButton(button);
+        return actionObject;
+    }
+
+    /**
+     * Follows the 'process pointer move action' algorithm in 17.3
+     * @param action Action being processed
+     * @param inputSourceType Source type
+     * @param id ID of input source that it's part of
+     * @param index Index within the 'actions' array
+     * @return Processed action object
+     * @throws InvalidArgumentException If failed to process, throw this. Means that args are bad.
+     */
+    public static ActionObject processPointerMoveAction(Action action, InputSourceType inputSourceType, String id, int index) throws InvalidArgumentException {
+        ActionObject actionObject = new ActionObject(id, inputSourceType, action.getType(), index);
+
+        // 1-3 Add the duration
+        Long duration = action.getDuration();
+        assertNullOrPositive(index, id, "duration", duration);
+        actionObject.setDuration(duration);
+
+        // 4-7 Add the origin
+        String origin = action.getOrigin();
```

Comment:
I'm not sure if origin is parsed properly if an element id is set there (like `{"element-6066-11e4-a52e-4f735466cecf": origin.id}`)

Reply:
Yeah, I just took the raw string. If we were only parsing W3C I would look for the `element-` prefix but I think we need to support MJSONWP as well.

Reply:
Aha, never mind I just thought it over again.... we need to expect the elements to be in the format `element-<element-id>`

Reply:
not really. The `6066-11e4-a52e-4f735466cecf` suffix seems to be constant. See https://github.com/SeleniumHQ/selenium/commit/d5b792d1c0e6dd164512819854c462dd30c75a19 or https://github.com/SeleniumHQ/selenium/commit/573baceab793fff51f228aee400d2f7f36987671

Reply:
Right, I got that confused with element id's....

So the format supposed to be: `{"origin": {"element-6066-11e4-a52e-4f735466cecf": "<element-uuid>}`?

Reply:
yes

Reply:
Alright, I updated it so that instead of 'Origin' being a string type I made an 'Origin' class that has a `type` property and an `elementId` property and made the JSON validator check if it's an object with that code, set the type to ELEMENT and set the elementId to the <element-uuid>. Otherwise, if it's a string, set the type and don't set elementId